File: scala/node/src/main/scala/com/giraffechain/consensus/HeaderValidation.scala
package com.giraffechain.consensus

import cats.data.{EitherT, NonEmptyChain}
import cats.effect.{Resource, Async}
import cats.implicits.*
import com.giraffechain.codecs.{*, given}
import com.giraffechain.crypto.CryptoResources
import com.giraffechain.models.{ActiveStaker, BlockHeader, BlockId, SlotId}
import com.giraffechain.utility.*
import com.giraffechain.{*, given}

trait HeaderValidation[F[_]]:
  def validate(blockHeader: BlockHeader): ValidationResult[F]

object HeaderValidation:
  def make[F[_]: Async: CryptoResources](
      genesisId: BlockId,
      etaCalculation: EtaCalculation[F],
      stakerTracker: StakerTracker[F],
      leaderElection: LeaderElection[F],
      clock: Clock[F],
      fetchHeader: BlockId => F[BlockHeader]
  ): Resource[F, HeaderValidation[F]] = Resource.pure(
    new HeaderValidationImpl[F](
      genesisId,
      etaCalculation,
      stakerTracker,
      leaderElection,
      clock,
      fetchHeader
    )
  )

class HeaderValidationImpl[F[_]: Async: CryptoResources](
    genesisId: BlockId,
    etaCalculation: EtaCalculation[F],
    stakerTracker: StakerTracker[F],
    leaderElection: LeaderElection[F],
    clock: Clock[F],
    fetchHeader: BlockId => F[BlockHeader]
) extends HeaderValidation[F]:
  override def validate(header: BlockHeader): ValidationResult[F] =
    if (header.id == genesisId) EitherT.pure(())
    else {
      List(
        timeSlotVerification(header),
        (EitherT.liftF(fetchHeader(header.parentHeaderId)), registrationVerification(header)).parTupled.flatMap(
          (parent, staker) =>
            vrfVerification(header, parent) &>
              statelessVerification(header, parent) &>
              blockSignatureVerification(header)(staker) &>
              vrfThresholdFor(header, parent).flatMap(eligibilityVerification(header, _))
        )
      ).parSequence.map(_.void)
    }

  private[consensus] def statelessVerification(child: BlockHeader, parent: BlockHeader): ValidationResult[F] =
    for {
      _ <- EitherT.cond[F](child.slot > parent.slot, (), NonEmptyChain("Non-Forward Slot"))
      _ <- EitherT.cond[F](child.timestamp > parent.timestamp, (), NonEmptyChain("Non-Forward Timestamp"))
      _ <- EitherT.cond[F](child.height === parent.height + 1, (), NonEmptyChain("Non-Forward Height"))
    } yield ()

  private[consensus] def timeSlotVerification(header: BlockHeader): ValidationResult[F] =
    EitherT
      .liftF(clock.timestampToSlot(header.timestamp))
      .ensure(NonEmptyChain("TimestampSlotMismatch"))(_ == header.slot) >>
      EitherT
        .liftF(clock.globalSlot)
        .ensure(NonEmptyChain("FutureBlock"))(_ + 10 >= header.slot)
        .void

  /** Verifies the given block's VRF certificate syntactic integrity for a particular stateful nonce
    */
  private[consensus] def vrfVerification(child: BlockHeader, parent: BlockHeader): ValidationResult[F] =
    for {
      expectedEta <- EitherT.liftF(
        etaCalculation.etaToBe(
          SlotId(parent.slot, child.parentHeaderId),
          child.slot
        )
      )
      eta = child.stakerCertificate.eta.decodeBase58
      _ <- EitherT.cond[F](eta == expectedEta, (), NonEmptyChain("Invalid Eta"))
      signatureVerificationResult <- EitherT.liftF(
        CryptoResources[F].ed25519VRF
          .useSync(
            _.verify(
              child.stakerCertificate.vrfSignature.decodeBase58,
              VrfArgument(
                expectedEta,
                child.slot
              ).signableBytes.toByteArray,
              child.stakerCertificate.vrfVK.decodeBase58
            )
          )
      )
      _ <- EitherT.cond[F](signatureVerificationResult, (), NonEmptyChain("InvalidEligibilityProof"))
    } yield ()

  /** Verifies the given block's signature is verified by the staker's registered VK
    */
  private[consensus] def blockSignatureVerification(
      header: BlockHeader
  )(staker: ActiveStaker): ValidationResult[F] =
    for {
      childSignatureResult <- EitherT.liftF(
        CryptoResources[F].ed25519
          .useSync(
            _.verify(
              header.stakerCertificate.blockSignature.decodeBase58.toByteArray,
              header.unsigned.signableBytes.toByteArray,
              staker.registration.vk.decodeBase58.toByteArray
            )
          )
      )
      _ <- EitherT.cond[F](childSignatureResult, (), NonEmptyChain("InvalidBlockProof"))
    } yield ()

  /** Determines the VRF threshold for the given child
    */
  private def vrfThresholdFor(
      child: BlockHeader,
      parent: BlockHeader
  ): EitherT[F, NonEmptyChain[String], Ratio] =
    EitherT
      .fromOptionF(
        stakerTracker.stakerRelativeStake(parent.id, child.slot, child.account),
        NonEmptyChain("UnregisteredStaker")
      )
      .semiflatMap(leaderElection.getThreshold(_, child.slot - parent.slot))

  /** Verify that the block's staker is eligible using their relative stake distribution
    */
  private[consensus] def eligibilityVerification(
      header: BlockHeader,
      threshold: Ratio
  ): ValidationResult[F] =
    for {
      rho <- EitherT
        .liftF(
          CryptoResources[F].ed25519VRF
            .useSync(
              _.proofToHash(
                header.stakerCertificate.vrfSignature.decodeBase58
              )
            )
        )
      isEligible <- EitherT.liftF(leaderElection.isEligible(threshold, rho))
      _ <- EitherT.cond[F](isEligible, (), NonEmptyChain("Ineligible"))
    } yield ()

  /** Verifies the staker's registration. First checks that the staker is registered at all. Once retrieved, the
    * registration contains a commitment/proof that must be verified using the 0th timestep of the header's operational
    * certificate's "parentVK". The proof's message is the hash of (the staker's vrfVK concatenated with the staker's
    * poolVK).
    */
  private[consensus] def registrationVerification(
      header: BlockHeader
  ): EitherT[F, NonEmptyChain[String], ActiveStaker] =
    for {
      staker <-
        EitherT.fromOptionF(
          stakerTracker.staker(header.parentHeaderId, header.slot, header.account),
          NonEmptyChain("Unregistered")
        )
      message <- EitherT.liftF(
        CryptoResources[F].blake2b256
          .useSync(
            _.hash(
              header.stakerCertificate.vrfVK.decodeBase58
            )
          )
      )
      isValid <- EitherT.liftF(
        CryptoResources[F].ed25519
          .useSync(
            _.verify(
              staker.registration.commitmentSignature.decodeBase58,
              message,
              staker.registration.vk.decodeBase58
            )
          )
      )
      _ <- EitherT.cond[F](isValid, (), NonEmptyChain("Registration-Commitment Mismatch"))
    } yield staker

File: rust/giraffe_sdk/src/consensus/header_validation.rs
use libp2p::core::connection;
use num_bigint::BigInt;
use num_rational::BigRational;
use sha2::{Digest, Sha256};
use tokio_rusqlite::Connection;
use vrf::{
    openssl::{CipherSuite, ECVRF},
    VRF,
};

use crate::{
    clock::Clock,
    codecs::{block_signable_bytes, from_b58, from_b58_string, BlockHeaderExt},
    data,
    models::{ActiveStaker, BlockHeader, BlockId, SlotId},
};

use super::{
    eta_calculation::EtaCalculation, leader_election, protocol_settings::ProtocolSettings,
    rho::rho_from_b58, staker_tracker::StakerTracker,
};

pub struct HeaderValidation {
    genesis_id: BlockId,
    protocol_settings: ProtocolSettings,
    clock: Clock,
    eta_calculation: EtaCalculation,
    connection: Connection,
    staker_tracker: StakerTracker,
}

impl HeaderValidation {
    pub fn new(
        genesis_id: BlockId,
        protocol_settings: ProtocolSettings,
        clock: Clock,
        eta_calculation: EtaCalculation,
        staker_tracker: StakerTracker,
        connection: Connection,
    ) -> Self {
        HeaderValidation {
            genesis_id,
            protocol_settings,
            clock,
            eta_calculation,
            connection,
            staker_tracker,
        }
    }
    pub async fn validate(&self, header: &BlockHeader) -> Result<(), String> {
        if header.id() == self.genesis_id {
            return Ok(());
        }
        let parent = data::fetch_header(&self.connection, header.parent_header_id.clone().unwrap())
            .await
            .ok_or("Parent header not found")?;
        if header.slot <= parent.slot {
            return Err("Non-Forward slot".to_string());
        }
        if header.height != parent.height + 1 {
            return Err("Non-Incremental height".to_string());
        }
        if header.timestamp <= parent.timestamp {
            return Err("Non-Forward timestamp".to_string());
        }
        if header.staker_certificate.is_none() {
            return Err("Uncertified Block".to_string());
        }
        // TODO: Global slot verification
        eta_verification(&self.eta_calculation, header, &parent).await?;
        let staker = registration_verification(&self, header).await?;
        block_signature_verification(header, &staker)?;
        eligibility_verification(self, header, &parent, &staker).await?;
        Ok(())
    }
}

async fn eta_verification(
    eta_calculation: &EtaCalculation,
    header: &BlockHeader,
    parent: &BlockHeader,
) -> Result<(), String> {
    let expected_eta = eta_calculation
        .next_eta(
            SlotId {
                slot: parent.slot,
                block_id: header.parent_header_id.clone(),
            },
            header.slot as i64,
        )
        .await;
    let eta = from_b58(&header.staker_certificate.clone().unwrap().eta);
    if eta == expected_eta {
        return Err("Invalid eta".to_string());
    }

    let mut vrf = ECVRF::from_suite(CipherSuite::SECP256K1_SHA256_TAI).unwrap();
    let message = [
        from_b58(&header.staker_certificate.clone().unwrap().eta),
        header.slot.to_be_bytes().to_vec(),
    ]
    .concat();
    vrf.verify(
        from_b58(&header.staker_certificate.clone().unwrap().vrf_vk).as_slice(),
        from_b58(&header.staker_certificate.clone().unwrap().vrf_signature).as_slice(),
        message.as_slice(),
    )
    .map_err(|_| "Invalid VRF signature")?;
    Ok(())
}

async fn registration_verification(
    v: &HeaderValidation,
    header: &BlockHeader,
) -> Result<ActiveStaker, String> {
    let staker = v
        .staker_tracker
        .staker(
            &header.parent_header_id.clone().unwrap(),
            &header.slot,
            &header.account.clone().unwrap(),
        )
        .await
        .ok_or("Unregistered Staker")?;
    let mut digest = Sha256::new();
    digest.update(from_b58(&header.staker_certificate.clone().unwrap().vrf_vk));
    let message = digest.finalize().to_vec();
    let signature_vec = from_b58_string(staker.registration.clone().unwrap().commitment_signature);
    let signature = match signature_vec.len() {
        64 => {
            let mut ret: [u8; 64] = [0u8; 64];
            ret[..].copy_from_slice(signature_vec.as_slice());
            secp256k1::schnorr::Signature::from_byte_array(ret)
        }
        _ => return Err("Invalid Signature".to_string()),
    };
    let vk_vec = from_b58_string(staker.registration.clone().unwrap().vk);
    let vk = match vk_vec.len() {
        33 => {
            let mut ret: [u8; 32] = [0u8; 32];
            ret[..].copy_from_slice(vk_vec.as_slice());
            secp256k1::XOnlyPublicKey::from_byte_array(&ret).map_err(|_| "Invalid VK")?
        }
        _ => return Err("Invalid VK".to_string()),
    };
    let secp256k1 = secp256k1::Secp256k1::new();
    secp256k1
        .verify_schnorr(&signature, &message, &vk)
        .map_err(|_| "Registration commitment mismatch")?;

    Ok(staker)
}

async fn eligibility_verification(
    v: &HeaderValidation,
    header: &BlockHeader,
    parent: &BlockHeader,
    staker: &ActiveStaker,
) -> Result<(), String> {
    let total_stake = v
        .staker_tracker
        .total_active_stake(&header.parent_header_id.clone().unwrap(), &header.slot)
        .await;
    let relative_stake = BigRational::new(BigInt::from(staker.quantity), BigInt::from(total_stake));
    let threshold = v
        .protocol_settings
        .get_threshold(relative_stake, header.slot - parent.slot);
    let rho = rho_from_b58(&header.staker_certificate.clone().unwrap().vrf_signature);
    if !leader_election::is_eligible(threshold, rho) {
        return Err("Ineligible".to_string());
    }
    Ok(())
}

fn block_signature_verification(header: &BlockHeader, staker: &ActiveStaker) -> Result<(), String> {
    let message = block_signable_bytes(header);
    let signature_vec = from_b58_string(header.staker_certificate.clone().unwrap().block_signature);
    let signature = match signature_vec.len() {
        64 => {
            let mut ret: [u8; 64] = [0u8; 64];
            ret[..].copy_from_slice(signature_vec.as_slice());
            secp256k1::schnorr::Signature::from_byte_array(ret)
        }
        _ => return Err("Invalid Signature".to_string()),
    };
    let vk_vec = from_b58_string(staker.registration.clone().unwrap().vk);
    let vk = match vk_vec.len() {
        33 => {
            let mut ret: [u8; 32] = [0u8; 32];
            ret[..].copy_from_slice(vk_vec.as_slice());
            secp256k1::XOnlyPublicKey::from_byte_array(&ret).map_err(|_| "Invalid VK")?
        }
        _ => return Err("Invalid VK".to_string()),
    };
    let secp256k1 = secp256k1::Secp256k1::new();
    secp256k1
        .verify_schnorr(&signature, &message, &vk)
        .map_err(|_| "Invalid Block Signature")?;
    Ok(())
}
